xde/src/route.rs

```diff
@@ -0,0 +1,514 @@
+        // from the physical address of the IP Lower Layer object
+        // member or the internet routing entry.
+        let src = (*ill).ill_phys_addr;
+        if src.is_null() {
+            opte::engine::dbg!(
+                "gateway ILL phys addr is NULL for {:?}",
+                gw_ip6
+            );
+            next_hop_probe(
+                ip6_dst,
+                Some(&gw_ip6),
+                EtherAddr::zero(),
+                EtherAddr::zero(),
+                b"gateway ILL phys addr is NULL\0",
+            );
+            return Err(underlay_dev);
+        }
+
+        let src: [u8; 6] = alloc::slice::from_raw_parts(src, 6)
+            .try_into()
+            .expect("src mac from pointer");
+
+        // Switch to the 2nd underlay device if we determine the source mac
+        // belongs to that device.
+        if src == ustate.u2.mac {
+            underlay_dev = &ustate.u2;
+        }
+
+        let src = EtherAddr::from(src);
+
+        // Step (4): Determine the destination address of the outer
+        // frame by retrieving the NCE entry for the gateway's
+        // link-local address.
+        let nce = DropRef::new(nce_refrele, ip::nce_lookup_v6(ill, &gw));
+        if nce.inner().is_null() {
+            opte::engine::dbg!("no NCE for gateway {:?}", gw_ip6);
+            next_hop_probe(
+                ip6_dst,
+                Some(&gw_ip6),
+                src,
+                EtherAddr::zero(),
+                b"no NCE for gateway\0",
+            );
+            return Err(underlay_dev);
+        }
+
+        let nce_common = (*nce.inner()).nce_common;
+        if nce_common.is_null() {
+            opte::engine::dbg!("no NCE common for gateway {:?}", gw_ip6);
+            next_hop_probe(
+                ip6_dst,
+                Some(&gw_ip6),
+                src,
+                EtherAddr::zero(),
+                b"no NCE common for gateway\0",
+            );
+            return Err(underlay_dev);
+        }
+
+        let mac = (*nce_common).ncec_lladdr;
+        if mac.is_null() {
+            opte::engine::dbg!("NCE MAC address is NULL {:?}", gw_ip6);
+            next_hop_probe(
+                ip6_dst,
+                Some(&gw_ip6),
+                src,
+                EtherAddr::zero(),
+                b"NCE MAC address if NULL for gateway\0",
+            );
+            return Err(underlay_dev);
+        }
+
+        let maclen = (*nce_common).ncec_lladdr_length;
+        assert!(maclen == 6);
+
+        let dst: [u8; 6] = alloc::slice::from_raw_parts(mac, 6)
+            .try_into()
+            .expect("mac from pointer");
+        let dst = EtherAddr::from(dst);
+
+        next_hop_probe(ip6_dst, Some(&gw_ip6), src, dst, b"\0");
+
+        Ok(Route { src, dst, underlay_dev })
+    }
+}
+
+/// A simple caching layer over `next_hop`.
+#[derive(Clone)]
+pub struct RouteCache(Arc<KRwLock<BTreeMap<RouteKey, CachedRoute>>>);
```

Comment:
I'll admit that I hadn't -- I was following the OPTE preference for `BTreeMap`s. But I threw something together quickly there in criterion with `RouteKey` and `CachedRoute` to insert/lookup one row into an arbitrary sized table:

```bash
   Compiling opte-bench v0.1.0 (/Users/kyle/gits/opte/bench)
    Finished bench [optimized + debuginfo] target(s) in 2.92s
     Running benches/userland.rs (target/release/deps/userland-57e28d11e88fdd6e)

mapquest/btreemap/get/0 time:   [287.02 ps 287.61 ps 288.30 ps]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe
mapquest/btreemap/get/2 time:   [3.8893 ns 3.9047 ns 3.9191 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
mapquest/btreemap/get/8 time:   [8.9178 ns 8.9306 ns 8.9465 ns]
Found 19 outliers among 100 measurements (19.00%)
  5 (5.00%) high mild
  14 (14.00%) high severe
mapquest/btreemap/get/64
                        time:   [8.1967 ns 8.2355 ns 8.2792 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
mapquest/btreemap/get/512
                        time:   [7.2225 ns 7.2445 ns 7.2702 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe
mapquest/btreemap/get/8192
                        time:   [2.8060 ns 2.8106 ns 2.8155 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild

mapquest/btreemap/insert/0
                        time:   [53.993 ns 54.513 ns 55.530 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe
mapquest/btreemap/insert/2
                        time:   [47.196 ns 47.271 ns 47.339 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
mapquest/btreemap/insert/8
                        time:   [57.601 ns 57.679 ns 57.760 ns]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
mapquest/btreemap/insert/64
                        time:   [290.32 ns 290.57 ns 290.83 ns]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) low mild
  4 (4.00%) high mild
mapquest/btreemap/insert/512
                        time:   [2.4069 µs 2.4108 µs 2.4145 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
mapquest/btreemap/insert/8192
                        time:   [50.270 µs 50.590 µs 50.858 µs]




mapquest/hashmap/get/0  time:   [287.28 ps 288.20 ps 289.29 ps]
Found 17 outliers among 100 measurements (17.00%)
  5 (5.00%) high mild
  12 (12.00%) high severe
mapquest/hashmap/get/2  time:   [22.259 ns 23.395 ns 24.472 ns]
mapquest/hashmap/get/8  time:   [21.733 ns 22.932 ns 24.172 ns]
mapquest/hashmap/get/64 time:   [22.045 ns 23.297 ns 24.493 ns]
mapquest/hashmap/get/512
                        time:   [19.729 ns 20.846 ns 22.009 ns]
mapquest/hashmap/get/8192
                        time:   [19.731 ns 20.825 ns 21.964 ns]

mapquest/hashmap/insert/0
                        time:   [46.374 ns 46.581 ns 46.796 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
mapquest/hashmap/insert/2
                        time:   [43.214 ns 43.935 ns 44.742 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
mapquest/hashmap/insert/8
                        time:   [66.543 ns 66.619 ns 66.698 ns]
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
mapquest/hashmap/insert/64
                        time:   [59.119 ns 60.854 ns 62.838 ns]
mapquest/hashmap/insert/512
                        time:   [295.78 ns 301.83 ns 308.87 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
mapquest/hashmap/insert/8192
                        time:   [231.58 ns 232.38 ns 233.19 ns]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low severe
  1 (1.00%) high mild
  1 (1.00%) high severe
```

This sort of tracks with my understanding of why we're using `BTreeMap`s in general throughout OPTE: faster lookup at the cost of more expensive insert.

Reply:
Perhaps a more obvious reason why we use `BTree{Map,Set}` that totally slipped my mind: there is no `HashMap` in `alloc`/`no_std`. So we would need to find and measure a suitable replacement if we wanted to go down that road. In the meantime I think the best call might be to limit the cache size.

Reply:
Sorry I don't have a ton of context here, but do we expect the nexthop cache to have a lot of regular churn?

Reply:
There shouldn't be in practice -- the above is very much worst-case, and I'm thinking about the case where a guest might have many active flows that are actively refreshing their routes and then add one (/n) more.

In the general case of a steady number of flows:
* An extant flow will call `next_hop` every 100ms in case preferences or reachability change. If there's an existing entry (even if expired), we should only be paying for lookup rather than a full insert on the `BTreeMap`.
* Expired routes are cleared out from the map every 1s.

The last part might add to the risk of churn; I'll change it up to expire flows which have expired by longer than the cleanup period to make refresh more consistent.

Reply:
My intuition here is that we want to optimize for queries rather than inserts. And if that intuition is not correct I really want to understand why. If this can be modeled as an LPM query, then using [poptrie](https://github.com/oxidecomputer/poptrie) might be a win.